### ISSUE TITLE
Dangers dependencies do not respect no_proxy environment variable

### ISSUE THREAD

**Issue by nikhilsh** [danger/danger#830]
# Report

## What did you do?
1) Set `no_proxy` environment variable
2) Set `DANGER_GITHUB_API_BASE_URL`
3) Set `DANGER_GITHUB_HOST`
4) Set `DANGER_GITHUB_API_TOKEN`
5) Run `bundle exec danger local`

## What did you expect to happen?
Expected danger to run locally

## What happened instead?
This was printed out:
`no_proxy is unsupported`
```/Users/xxx/.rvm/rubies/ruby-2.4.0/lib/ruby/2.4.0/net/http/response.rb:122:in `error!': 503 "Service Unavailable" (Faraday::ConnectionFailed)```
## Your Environment

* Which CI are you running on? Local
* Are you running the latest version of Danger? Yes
* What is your Dangerfile?

  ```ruby
    # PR is a work in progress and shouldn't be merged yet
    warn "PR is classed as Work in Progress" if github.pr_title.include? "[WIP]"

    # Warn when there is a big PR
    warn "Big PR, consider splitting into smaller" if git.lines_of_code > 500

    # Ensure a clean commits history
    if git.commits.any? { |c| c.message =~ /^Merge branch '#{github.branch_for_base}'/ }
      fail "Please rebase to get rid of the merge commits in this PR"
    end

    # Mainly to encourage writing up some reasoning about the PR, rather than
    # just leaving a title
    if github.pr_body.length < 5
      fail "Please provide a summary in the Pull Request description"
    end

    # If these are all empty something has gone wrong, better to raise it in a comment
    if git.modified_files.empty? && git.added_files.empty? && git.deleted_files.empty?
      fail "This PR has no changes at all, this is likely an issue during development."
    end

    has_app_changes = !git.modified_files.grep(/ProjectName/).empty?
    has_test_changes = !git.modified_files.grep(/ProjectNameTests/).empty?

    # If changes are more than 10 lines of code, tests need to be updated too
    if has_app_changes && !has_test_changes && git.lines_of_code > 10
      fail("Tests were not updated", sticky: false)
    end

    # Info.plist file shouldn't change often. Leave warning if it changes.
    is_plist_change = git.modified_files.sort == ["ProjectName/Info.plist"].sort

    if !is_plist_change
      warn "Plist changed, don't forget to localize your plist values"
    end

    podfile_updated = !git.modified_files.grep(/Podfile/).empty?

    # Leave warning, if Podfile changes
    if podfile_updated
      warn "The `Podfile` was updated"
    end

    # This is swiftlint plugin. More info: https://github.com/ashfurrow/danger-swiftlint
    #
    # This lints all Swift files and leave comments in PR if 
    # there is any issue with linting
    swiftlint.lint_files
    swiftlint.lint_files inline_mode: true
  ```

**Comment by orta** [comment]
Looks like support was added pretty recently, https://github.com/lostisland/faraday/pull/658 want to look at updating the dependency?

**Comment by nikhilsh** [comment]
I've already tried setting `faraday` to `0.12.1` but it didn't seem to help

**Comment by nikhilsh** [comment]
@orta Does this look like a possible reason as to why `no_proxy` is not being respected now? https://github.com/lostisland/faraday/issues/701. Basically calling Faraday.get without first initialising the connection. If so, this should be fixed in `0.12.2`

**Comment by orta** [comment]
Yeah.- I could imagine that being it

**Comment by nikhilsh** [comment]
Hi @orta, I've looked into `faraday 0.12.2` which supposedly would have fixed the `no_proxy` issue, but seems like we're getting two other issues, that I've figured out locally, but would love your opinion before submitting a PR.

The first error is: 
` 'connect_nonblock': SSL_connect returned=1 errno=0 state=error: certificate verify failed (Faraday::SSLError)`

This fails as our GitHub Enterprise has a self-signed SSL cert and hence the verification fails when dealing with Octokit. I've managed to get `bundle exec danger pr` to work by adding a flag `--verify-ssl` flag and within `pr.rb`, I've modified the code as such:

```ruby
def configure_octokit(cache_dir)
  # setup caching for Github calls to hitting the API rate limit too quickly
  cache_file = File.join(cache_dir, "danger_local_cache")
  cache = HTTPCache.new(cache_file, clear_cache: @clear_http_cache)
  p @verify_ssl
  Octokit.configure do |config|
    config.connection_options[:ssl] = { :verify => @verify_ssl }
  end
  Octokit.middleware = Faraday::RackBuilder.new do |builder|
    builder.use Faraday::HttpCache, store: cache, serializer: Marshal, shared_cache: false
    builder.use Octokit::Response::RaiseError
    builder.adapter Faraday.default_adapter
  end
end
```

This works in solving the SSL issue, but I'm not too sure if this is the only/best place to set this. Would love your advice on this.

The second issue seems to be https://github.com/bundler/bundler/issues/5300, and the fix for that is by installing a temporary gem: http://ermaker.github.io/blog/2017/02/06/proxy-error-on-ruby-2.4.0.html. This error happens when a no_proxy variable is set on Ruby version 2.4.0/2.4.1. I could submit another PR for that but not too sure if you would like to install that gem. 

Thank you!

**Comment by orta** [comment]
Him, I think that seems like a reasonable tradeoff - TBH, I don't know too much about all the HTTP caching, it came in as a PR 🍡 

I'm happy to have it skipping verification, as it's not really a potential attack vector for anyone.

WRT the gem, that seems  fine too

**Comment by nikhilsh** [comment]
Fixed in #880. Closing this issue. Thanks @orta!
